src/org/zaproxy/zap/extension/ascanrulesBeta/InsecureHTTPMethod.java

```diff
@@ -1,455 +1,702 @@
+        // in as simple as possible a manner.
+        Socket socket = null;
+        try {
+            ProxyClient client = new ProxyClient();
+            client.getHostConfiguration().setProxy(connecthost, connectport);
+            client.getHostConfiguration().setHost(thirdpartyHost, thirdpartyPort);
+            ConnectResponse connectResponse = client.connect();
+            StatusLine statusLine = connectResponse.getConnectMethod().getStatusLine();
+
+            if (log.isDebugEnabled())
+                log.debug("The status line returned: " + statusLine);
+
+            int statusCode = statusLine.getStatusCode();
+            socket = connectResponse.getSocket();
+            if (socket != null && statusCode == HttpStatus.SC_OK) {
+                // we have a socket and a 200 status.
+                // Could still be a false positive though, if the server ignored
+                // the method,
+                // and did not recognise the URL, so redirected to a login page,
+                // for instance
+                // Remediation: Check the contents match the expected third
+                // party contents.
+                if (log.isDebugEnabled())
+                    log.debug("Raw Socket established, in theory to " + thirdpartyHost);
+
+                OutputStream os = socket.getOutputStream();
+                InputStream is = socket.getInputStream();
+
+                PrintWriter pw = new PrintWriter(os, false);
+                pw.write("GET http://" + thirdpartyHost + ":" + thirdpartyPort + "/ HTTP/1.1\n");
+                pw.write("Host: " + thirdpartyHost + "\n\n");
+                pw.flush();
+
+                // read the response via a 4k buffer
+                ByteArrayOutputStream bos = new ByteArrayOutputStream();
+                byte[] buffer = new byte[1024 * 4];
+                int bytesRead = is.read(buffer);
+                int totalBytesRead = 0;
+                while (bytesRead > -1) {
+                    totalBytesRead += bytesRead;
+                    bos.write(buffer, 0, bytesRead);
+                    bytesRead = is.read(buffer);
+                }
+                String response = new String(bos.toByteArray());
+                if (log.isDebugEnabled())
+                    log.debug("Response is " + totalBytesRead + " bytes: \n" + response);
+
+                Matcher m = thirdPartyContentPattern.matcher(response);
+                if (m.matches()) {
+                    log.debug("Response matches expected third party pattern!");
+                    is.close();
+                    os.close();
+                    bos.close();
+                    socket.close();
+                    return true;
+                } else {
+                    log.debug("Response does *not* match expected third party pattern");
+                }
+                is.close();
+                os.close();
+                bos.close();
+                socket.close();

```

Comment:
Instead of duplicating all these close operations they could be done in a finally block?